Add github token action parameter

antifuchs opened the pull request:
This PR adds a `github_access_token` parameter to the action, which then sets that token up as an `extra-access-tokens` entry for github.com. If no token is given, the action now configures the GHA workflow run's token, which allows a greater rate limit than unauthenticated operation.

Allowing custom tokens allows even greater rate limits, and access to private repos that the token has access to.

antifuchs committed:
Add github token action parameter

Setting a github token allows git operations to run with more permissive rate limits (allowing flakes to function better) and to optionally access private repos.

antifuchs commented on do-the-lix-thing:
```diff
@@ -129,7 +129,16 @@ set -x
-	< /etc/nix/nix.conf sed -e 's/nixpkgs=flake:nixpkgs/nixpkgs=channel:nixos-unstable/'
+	if [ -z "${INPUT_GITHUB_ACCESS_TOKEN:-}" ]; then
+		echo "Configuring default github access token and overriding default nixpkgs with channel-based config" >&2
+		< /etc/nix/nix.conf sed -e 's/nixpkgs=flake:nixpkgs/nixpkgs=channel:nixos-unstable/'
```
I'm not sure that with the elevated rate limits, this channel override would be necessary at all anymore. What do you think?

samueldr replied:
Hmmm... I'm on the fence.

On one hand, it's [likely] true that it's not needed anymore.

On the other hand, having different "default" behaviours is an issue. The least it differs, the least likely it introduces harder to debug problems.

Now, to make things just this much more fun to reason about: changing to the NixOS channel release mechanisms ***is*** a change in behaviour, compared to (apparently) the default Lix assumptions.

The change was made for practical purposes, to work around an issue within this CI environment... But I have *opinions* (that I haven't taken time to discuss about) about this. Though that should be discussed with upstream.

So that leaves these things to consider:

 - (A) Should the action even change this behaviour by default?
    - I think so, because of how it breaks by default, but... the change with the tokens may or may not work around this issue enough for that breakage
 - (B) Should a discrete option be made to apply this fix?
    - Probably? To work around the breakage by default.
    - If so, applied by default I think, all the time.
    - I don't think this could cause any issue with Flakes proper, as `NIX_PATH` (which is what this is) is not used with Flakes
    - My *additional* personal take on this, is that since Flakes are an experimental feature, still, that the defaults for builds without Flakes, which is what `NIX_PATH` effectively is used for, shouldn't involve Flakes.

* * *

<details><summary>(thoughts detailed...)</summary>

My to-be-discussed-upstream concern:

I'm not so sure about using a `flake` input for Nixpkgs, it's weird to end-up relying on a, albeit the upstream, github-hosted repository, rather than the published tarballs, for Nixpkgs. Even within the fully-flake-accepting model, I'm not sure this is a good outcome.

And I'm not even yet considering "actual" flake inputs in this thought...

It's ***really weirding me out***, ending-up relying on some third party service that can and does get limited, to access the package set, which has always been, and still is self-hosted and self-published in a much less limiting manner.

</details>

antifuchs replied:
Hm, yeah, the way you explain it makes sense & now I agree that the action should substitute the channel for the flake default.

Given this, I'll change the channel unconditionally again & maybe we can take this discussion to the lix devs (and maybe introduce a more generalized setting for people who need the flake channel-like behavior... I do not (-:)

antifuchs commented:
Updated - please take another look (:

samueldr commented:
:thinking:

> An action can access the GITHUB_TOKEN through the github.token context even if the workflow does not explicitly pass the GITHUB_TOKEN to the action. As a good security practice, you should always make sure that actions only have the minimum access they require by limiting the permissions granted to the GITHUB_TOKEN. For more information, see [Permissions for the GITHUB_TOKEN](https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#permissions-for-the-github_token).

https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow

I don't think that it means that it's always making `github.token` available. And uh, see the workflow runs :see_no_evil:.

antifuchs commented:
Hm, the way I read this is that the `github.token` can be accessed by actions run by forks! The CI failures point at line 5, which has a `${{github.token}}` reference in the description of a parameter... wonder if github is doing a recursion there. I'll un-interpolate that and add some quotes to stuff to be safer.

antifuchs commented:
Seems that was it, what do you think?

samueldr commented:
Yes, that was it.

Sorry, this completely slipped my grasp.

samueldr committed:
support: Import update-readme to project

With three projects using it, it might be time to consider breaking it
out into a sort of gha-utils...

samueldr committed:
CI: Add check for README docs

samueldr committed:
README: Add template parts for updater script

This commit will fail the check by design.

samueldr committed:
README: Update with support/update-readme.rb

samueldr commented:
:tada: https://github.com/samueldr/lix-gha-installer-action/actions/runs/13558841791/job/37898237900

samueldr commented:
Thank you!
